File: android/core/model/src/main/kotlin/com/mycorrhizal/crm/model/network/OccasionEvents.kt
package com.mycorrhizal.crm.model.network

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * A one-off event the user is hosting (docs/adrs/0026-occasions-events.md,
 * issue #1228) — the Android parity slice of the web Occasions events surface.
 * Unlike the standing [OccasionObligation] rule, an event is a concrete
 * instant (`starts_at`/`ends_at` are RFC 3339) with an invitee/RSVP ledger.
 */
@JsonClass(generateAdapter = true)
data class OccasionEvent(
    val id: String = "",
    @Json(name = "created_at") val createdAt: String? = null,
    @Json(name = "updated_at") val updatedAt: String? = null,
    val title: String = "",
    @Json(name = "starts_at") val startsAt: String = "",
    @Json(name = "ends_at") val endsAt: String? = null,
    val location: String? = null,
    val sensitivity: String = "normal",
    val notes: String? = null,
    /** Change-feed tombstone marker (T17); only present via `?since=`. */
    val deleted: Boolean = false,
)

/** POST/PUT /occasion-events request body. */
@JsonClass(generateAdapter = true)
data class OccasionEventInput(
    val title: String,
    @Json(name = "starts_at") val startsAt: String,
    @Json(name = "ends_at") val endsAt: String? = null,
    val location: String? = null,
    val sensitivity: String = "normal",
    val notes: String? = null,
)

/** GET /occasion-events — cursor-paginated `{ occasion_events, total, next_cursor, limit }`. */
@JsonClass(generateAdapter = true)
data class OccasionEventsResponse(
    @Json(name = "occasion_events") val occasionEvents: List<OccasionEvent> = emptyList(),
    val total: Int = 0,
    @Json(name = "next_cursor") val nextCursor: String? = null,
    val limit: Int = 0,
)

/** POST /occasion-events response — wrapped `{ occasion_event }`. */
@JsonClass(generateAdapter = true)
data class CreateOccasionEventResponse(
    @Json(name = "occasion_event") val occasionEvent: OccasionEvent? = null,
)

/** One invitee/RSVP join row (raw). */
@JsonClass(generateAdapter = true)
data class OccasionEventAttendee(
    val id: String = "",
    @Json(name = "event_id") val eventId: String = "",
    @Json(name = "entity_id") val entityId: String = "",
    val rsvp: String = "pending",
)

/**
 * An attendee enriched with its contact's display data, returned by
 * GET /occasion-events/{id}. `contactId` is the numeric Contact id a client
 * links to; `entityId` is the invited Contact.VCardUID.
 */
@JsonClass(generateAdapter = true)
data class OccasionEventAttendeeView(
    val id: String = "",
    @Json(name = "event_id") val eventId: String = "",
    @Json(name = "entity_id") val entityId: String = "",
    @Json(name = "contact_id") val contactId: Int = 0,
    @Json(name = "contact_name") val contactName: String = "",
    val rsvp: String = "pending",
)

/** GET /occasion-events/{id} — `{ occasion_event, attendees }`. */
@JsonClass(generateAdapter = true)
data class OccasionEventDetail(
    @Json(name = "occasion_event") val occasionEvent: OccasionEvent? = null,
    val attendees: List<OccasionEventAttendeeView> = emptyList(),
)

/** POST /occasion-events/{id}/attendees request body. */
@JsonClass(generateAdapter = true)
data class OccasionEventAttendeeInput(
    @Json(name = "entity_id") val entityId: String,
    val rsvp: String? = null,
)

/** POST /occasion-events/{id}/attendees response — wrapped `{ attendee }`. */
@JsonClass(generateAdapter = true)
data class AddOccasionEventAttendeeResponse(
    val attendee: OccasionEventAttendee? = null,
)

/** PUT /occasion-events/{id}/attendees/{vcard_uid} request body. */
@JsonClass(generateAdapter = true)
data class OccasionEventAttendeeUpdateInput(
    val rsvp: String,
)

/** One candidate from GET /occasion-events/invitee-suggestions. */
@JsonClass(generateAdapter = true)
data class InviteeSuggestion(
    @Json(name = "contact_id") val contactId: Int = 0,
    @Json(name = "contact_name") val contactName: String = "",
    @Json(name = "entity_id") val entityId: String = "",
)

/** GET /occasion-events/invitee-suggestions response. */
@JsonClass(generateAdapter = true)
data class InviteeSuggestionsResponse(
    val suggestions: List<InviteeSuggestion> = emptyList(),
)

File: frontend/src/api/occasionEvents.ts
// OccasionEvent API calls (docs/adrs/0026-occasions-events.md, issue #1228): a
// one-off event the user is hosting, plus its attendee/RSVP list. RSVP is a
// manually-recorded status ("what they told you"), not a delivered invitation —
// this feature sends nothing.
import { API_BASE_URL, apiFetch, getAuthHeaders, parseErrorResponse } from './client';
import type { OccasionSensitivity } from './occasionObligations';

export type OccasionEventRSVP = 'pending' | 'accepted' | 'declined' | 'maybe';

export const OCCASION_EVENT_RSVPS: OccasionEventRSVP[] = [
  'pending',
  'accepted',
  'declined',
  'maybe',
];

export interface OccasionEvent {
  id: string;
  created_at: string;
  updated_at: string;
  title: string;
  starts_at: string;
  ends_at?: string;
  location?: string;
  sensitivity: OccasionSensitivity;
  notes?: string;
  // Change-feed tombstone marker (T17), only true via ?since=.
  deleted?: boolean;
}

export interface OccasionEventInput {
  title: string;
  starts_at: string;
  ends_at?: string | null;
  location?: string;
  sensitivity?: OccasionSensitivity;
  notes?: string;
}

export interface OccasionEventAttendee {
  id: string;
  created_at: string;
  updated_at: string;
  event_id: string;
  entity_id: string;
  rsvp: OccasionEventRSVP;
}

export interface OccasionEventAttendeeView {
  id: string;
  event_id: string;
  entity_id: string;
  contact_id: number;
  contact_name: string;
  rsvp: OccasionEventRSVP;
}

export interface InviteeSuggestion {
  contact_id: number;
  contact_name: string;
  entity_id: string;
}

export interface OccasionEventsResponse {
  occasion_events: OccasionEvent[];
  total: number;
  next_cursor: string;
  limit: number;
}

export interface OccasionEventDetail {
  occasion_event: OccasionEvent;
  attendees: OccasionEventAttendeeView[];
}

// GET /occasion-events
export async function getOccasionEvents(params?: {
  from?: string;
  to?: string;
  cursor?: string;
  limit?: number;
}): Promise<OccasionEventsResponse> {
  const { from, to, cursor, limit = 100 } = params || {};
  const queryParams = new URLSearchParams({ limit: limit.toString() });
  if (from) queryParams.append('from', from);
  if (to) queryParams.append('to', to);
  if (cursor) queryParams.append('cursor', cursor);
  const response = await apiFetch(`${API_BASE_URL}/occasion-events?${queryParams.toString()}`, {
    headers: getAuthHeaders(),
  });
  if (!response.ok) throw await parseErrorResponse(response);
  return response.json();
}

// GET /occasion-events/:id
export async function getOccasionEvent(id: string): Promise<OccasionEventDetail> {
  const response = await apiFetch(`${API_BASE_URL}/occasion-events/${id}`, {
    headers: getAuthHeaders(),
  });
  if (!response.ok) throw await parseErrorResponse(response);
  return response.json();
}

// POST /occasion-events — create is wrapped in {occasion_event: ...}
export async function createOccasionEvent(input: OccasionEventInput): Promise<OccasionEvent> {
  const response = await apiFetch(`${API_BASE_URL}/occasion-events`, {
    method: 'POST',
    headers: getAuthHeaders(),
    body: JSON.stringify(input),
  });
  if (!response.ok) throw await parseErrorResponse(response);
  const result = await response.json();
  return result.occasion_event;
}

// PUT /occasion-events/:id (full-replace)
export async function updateOccasionEvent(
  id: string,
  input: OccasionEventInput,
): Promise<OccasionEvent> {
  const response = await apiFetch(`${API_BASE_URL}/occasion-events/${id}`, {
    method: 'PUT',
    headers: getAuthHeaders(),
    body: JSON.stringify(input),
  });
  if (!response.ok) throw await parseErrorResponse(response);
  return response.json();
}

// DELETE /occasion-events/:id (soft delete)
export async function deleteOccasionEvent(id: string): Promise<void> {
  const response = await apiFetch(`${API_BASE_URL}/occasion-events/${id}`, {
    method: 'DELETE',
    headers: getAuthHeaders(),
  });
  if (!response.ok) throw await parseErrorResponse(response);
}

// POST /occasion-events/:id/attendees
export async function addOccasionEventAttendee(
  eventId: string,
  input: { entity_id: string; rsvp?: OccasionEventRSVP },
): Promise<OccasionEventAttendee> {
  const response = await apiFetch(`${API_BASE_URL}/occasion-events/${eventId}/attendees`, {
    method: 'POST',
    headers: getAuthHeaders(),
    body: JSON.stringify(input),
  });
  if (!response.ok) throw await parseErrorResponse(response);
  const result = await response.json();
  return result.attendee;
}

// PUT /occasion-events/:id/attendees/:vcard_uid — records the RSVP status.
export async function updateOccasionEventAttendee(
  eventId: string,
  vcardUid: string,
  rsvp: OccasionEventRSVP,
): Promise<OccasionEventAttendee> {
  const response = await apiFetch(
    `${API_BASE_URL}/occasion-events/${eventId}/attendees/${vcardUid}`,
    {
      method: 'PUT',
      headers: getAuthHeaders(),
      body: JSON.stringify({ rsvp }),
    },
  );
  if (!response.ok) throw await parseErrorResponse(response);
  return response.json();
}

// DELETE /occasion-events/:id/attendees/:vcard_uid
export async function removeOccasionEventAttendee(
  eventId: string,
  vcardUid: string,
): Promise<void> {
  const response = await apiFetch(
    `${API_BASE_URL}/occasion-events/${eventId}/attendees/${vcardUid}`,
    {
      method: 'DELETE',
      headers: getAuthHeaders(),
    },
  );
  if (!response.ok) throw await parseErrorResponse(response);
}

// GET /occasion-events/invitee-suggestions — expands circles into candidates,
// excluding anyone already invited when eventId is given.
export async function getInviteeSuggestions(params: {
  circleIds: string[];
  eventId?: string;
}): Promise<InviteeSuggestion[]> {
  const queryParams = new URLSearchParams({ circle_ids: params.circleIds.join(',') });
  if (params.eventId) queryParams.append('event_id', params.eventId);
  const response = await apiFetch(
    `${API_BASE_URL}/occasion-events/invitee-suggestions?${queryParams.toString()}`,
    { headers: getAuthHeaders() },
  );
  if (!response.ok) throw await parseErrorResponse(response);
  const result = await response.json();
  return result.suggestions || [];
}
